File: Depth-First Search/Backtracking/Hard/LET_no37_Sudoku Solver.cpp
class Solution {
private:
    int rows[9][9], cols[9][9], blocks[9][9];
public:
    void solveSudoku(vector<vector<char>>& board) {
        memset(rows, 0, sizeof(rows));
        memset(cols, 0, sizeof(cols));
        memset(blocks, 0, sizeof(blocks));
        for (int i = 0; i < 9; i++) {
            for (int j = 0; j < 9; j++) {
                if (board[i][j] != '.') {
                    int num = board[i][j] - '1', k = i / 3 * 3 + j / 3;
                    rows[i][num] = cols[j][num] = blocks[k][num] = 1;
                }
            }
        }
        
        solver(board, 0, 0);
    }
    bool solver(vector<vector<char>>& board, int i, int j) {
        if (i == 9)
            return true;
        if (j == 9)
            return solver(board, i + 1, 0);
        if (board[i][j] != '.')
            return solver(board, i, j + 1);
        
        int k = i / 3 * 3 + j / 3;
        for (char c = '1'; c <= '9'; c++) {
            int num = c - '1';
            if (check(i, j, k, num)) {
                board[i][j] = c;
                rows[i][num] = cols[j][num] = blocks[k][num] = 1;
                if (solver(board, i, j+ 1))
                    return true;
                rows[i][num] = cols[j][num] = blocks[k][num] = 0;
                board[i][j] = '.';
            }
        }
        
        return false;
    }
    bool check(int i, int j, int k, int num) {
        if (rows[i][num] || cols[j][num] || blocks[k][num])
            return false;
        return true;
    }
};

File: Depth-First Search/Backtracking/Hard/LET_no37_Sudoku Solver.py
class Solution:
    def __init__(self):
        self.rows = [[False] * 9 for _ in range(9)]
        self.cols = [[False] * 9 for _ in range(9)]
        self.blocks = [[False] * 9 for _ in range(9)]
    
    def solveSudoku(self, board: List[List[str]]) -> None:
        """
        Do not return anything, modify board in-place instead.
        """
        for i in range(9):
            for j in range(9):
                if board[i][j] != '.':
                    num = int(board[i][j]) - 1
                    k = i // 3 * 3 + j // 3
                    self.rows[i][num] = self.cols[j][num] = self.blocks[k][num] = True
                    
        self.solver(board, 0, 0)
    
    def solver(self, board, i, j):
        if i == 9:
            return True
        if j == 9:
            return self.solver(board, i + 1, 0)
        if board[i][j] != '.':
            return self.solver(board, i, j + 1)
        
        k = i // 3 * 3 + j // 3
        for num in range(9):
            if self.check(i, j, k, num):
                self.rows[i][num] = self.cols[j][num] = self.blocks[k][num] = True
                board[i][j] = chr(ord('1') + num)
                if self.solver(board, i, j + 1):
                    return True
                board[i][j] = '.'
                self.rows[i][num] = self.cols[j][num] = self.blocks[k][num] = False
        
        return False
    
    def check(self, i, j, k, num):
        return not (self.rows[i][num] or self.cols[j][num] or self.blocks[k][num])